designs/comprehensions-design.md

```diff
@@ -0,0 +1,174 @@
+# Comprehensions Design Document
+
+---
+
+## Test Cases
+
+Comprehensions follow this format:
+```
+<expr> for <name> in <iterable-expr> [if <expr>]?
+```
+A comprehension expression could be placed in a pair of square brackets ```[]``` (list comprehension), a pair of curly brackets ```{}``` (set/dictionary comprehension), or a pair of parentheses ```()``` (generator).
+
+The expression before the ```for``` keyword could be any valid expression -- plain literals or free variables, unary/binary operations, function calls, class fields or method calls, or ternary expressions which we haven't implemented yet.
+
+The iterable object after the ```in``` keyword could be of various type -- ```list```, ```set```, ```dictionary```, ```tuple```, ```generator```, or even another comprehension expression.
+
+In addition, we can have an optional ```if``` condition to filter the items in the iterable object.
+
+Therefore, there are so many scenarios to write test cases for. Since this is the first milestone and many cases are dependent on the implementation of other groups (specifically, list, for-loop/iterator, string, set/tuple/dictionary, and built-in library), __we plan to only support a range of integers as the iterable object by the end of week 7__. However, we are looking into possible ways to implement an ```Iterable``` interface that supports ```next()``` and ```has_next()``` to facilitate iterating over the above types.
```

Comment:
This makes a lot of sense. I think you might want to consider suggesting some more tests below that work with this assumption, since many of these cannot run until other groups make some progress.

There are also several groups working with something like `range` to get started. It's worth figuring out how to coordinate the definition of `range` so it will work for all of you. Check out https://github.com/ucsd-cse231-s22/chocopy-wasm-compiler-B/pull/18#pullrequestreview-965474421, for instance.